services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,263 @@
+import org.keycloak.OAuthErrorException;
+import org.keycloak.authentication.AuthenticationProcessor;
+import org.keycloak.common.ClientConnection;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.AuthenticationFlowModel;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientScopeModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserConsentModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.models.utils.AuthenticationFlowResolver;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.protocol.ciba.utils.DecoupledAuthnResult;
+import org.keycloak.protocol.ciba.utils.DecoupledAuthnResultParser;
+import org.keycloak.protocol.ciba.utils.DecoupledAuthStatus;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.protocol.oidc.ext.OIDCExtProvider;
+import org.keycloak.protocol.oidc.utils.AuthorizeClientUtil;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.sessions.RootAuthenticationSessionModel;
+
+public abstract class DecoupledAuthenticationProviderBase implements DecoupledAuthenticationProvider, OIDCExtProvider {
+
+    private static final Logger logger = Logger.getLogger(DecoupledAuthenticationProviderBase.class);
+
+    protected KeycloakSession session;
+    protected EventBuilder event;
+
+    protected MultivaluedMap<String, String> formParams;
+
+    protected RealmModel realm;
+
+    protected ClientModel client;
+    protected Map<String, String> clientAuthAttributes;
+
+    @Context
+    protected HttpHeaders headers;
+    @Context
+    protected HttpRequest httpRequest;
+    @Context
+    protected HttpResponse httpResponse;
+    @Context
+    protected ClientConnection clientConnection;
+
+    protected Cors cors;
+
+    public DecoupledAuthenticationProviderBase(KeycloakSession session) {
+        this.session = session;
+        realm = session.getContext().getRealm();
+    }
+
+    @Override
+    public void setEvent(EventBuilder event) {
+        this.event = event;
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Path("/")
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response processDecoupledAuthnResult() {
+        event.event(EventType.LOGIN);
+        headers = session.getContext().getContextObject(HttpHeaders.class);
+        httpRequest = session.getContext().getContextObject(HttpRequest.class);
```

Comment:
Aren't those fields automatically injected as they are annotated with `@Context` ?

Reply:
@pedroigor I've missed that this `@Context` does not work so that these should be removed.
This class uses the endpoint extension and it does not inject context.
https://github.com/keycloak/keycloak/blob/750f5fdb0a8e9a87ec14f256d42f81ece74748d4/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java#L302-L310